ND Spline Evaluator

tretre91 opened the pull request:
This PR adds a SplineEvaluatorND class working in any number of dimensions. Currently, the template arguments to the class and to the methods are passed using `TypeSeq`s, and the aliases `alias_1, alias_2, etc` are replaced by templated equivalents `alias<0>, alias<1>, etc`.

The tests for this are just the 3D spline test and one of the 1D spline test modified to use the ND evaluator. We might want to factorize some of the test code or use the same test files with some `#ifdef`s to enable the ND or regular Evalutor

tretre91 committed:
Add ND evaluator
- works for host backends for the moment

tretre91 committed:
Remove a variadic template argument in one of the `deriv` overloads

tretre91 committed:
Change the tests after rebasing on main

tretre91 committed:
Remove most of the variadic template parameters
- test the 1d ND evaluator
- rename the 3d test file

tpadioleau reviewed (commented):
You also need to declare the new header file in the `target_sources` to make the installation test pass

tretre91 committed:
Add spline_evaluator_nd.hpp in the CMakeLists

tretre91 committed:
Make some variables const and add documentation

tretre91 committed:
Fix implicit capture of this warnings

tretre91 committed:
Remove the variadic template from the deriv_dim_I methods
- Make the DerivDims template argument non-variadic
- Add tests for one of the deriv_dim_I overloads

tretre91 committed:
Get rid of the index_sequence template argument

tretre91 committed:
Rename the 3D ND tests

tretre91 committed:
Fix typo

tretre91 committed:
Remove the deriv_dim_I method from the evaluator

tretre91 committed:
Update the deriv methods to use DiscreteElements

tretre91 committed:
Replace uses of for_each with the host and device versions

tretre91 committed:
Use cexa::tuple instead of std::tuple
- This will need to be cleaned, some headers (ex: apply.hpp) are not
  used

tretre91 committed:
Fix cmake format

tretre91 committed:
Remove extra semicolon and define a tuple copy contructor

tretre91 committed:
Apply the even degree fixes from #1004

tretre91 committed:
Add missing typename, remove ranges as it seems to cause an hipcc ice

tretre91 committed:
Replace a parameter pack in the tests to avoid cuda lambda restrictions

tpadioleau commented:
> The errors don't actually come from my changes it seems, on adastra mi250x in Release, I have failures on both the Nd and 1d/3d tests (kokkos/kokkos kernels 5.0.2, ginkgo 1.11.0, amdclang++ 6.2.1)
> All errors are related to uniform splines, greville bc (in degree 2 or 5), hermite bc (in degrees 5)

I think we need rocm 6.3.3 to make the splines tests work. I don't know what is happening for `SparseDiscreteDomainTest`.

tretre91 committed:
Fix som clang-tidy issues in the tuple snapshot

tretre91 committed:
Fix a typo in one of the tests

tretre91 committed:
Use an array instead of a tuple for the for_each indices

tretre91 committed:
Add licences for the tuple

tretre91 committed:
Update the tuple
- update the tuple to c++20
- fix clang-tidy warnings
- do not use CTAD (as nvcc fails with single argument ctors)

tretre91 committed:
Fix clang-tidy 21 warnings and use clang-format 22

tretre91 committed:
Use concepts and use class instead of typename for template parameters

tretre91 committed:
Fix some warnings and fix compilation with clang 16

tretre91 committed:
Remove redundant typename

tretre91 committed:
Remove an unused include in the tests

tretre91 committed:
Fix compilation of the tuple with nvcc 12.2

tpadioleau commented:
> The tests are passing on Jean-Zay h100

Thanks, ready to review then ?

tpadioleau commented on src/ddc/detail/tuple/impl/apply.hpp:
```diff
@@ -0,0 +1,159 @@
+// NOLINTBEGIN(readability-identifier-naming)
+// SPDX-FileCopyrightText: 2026 CExA-project
```
I don't see the copyrights of the initial implementation ?

Also I don't think we need the date, we don't put  it in the rest of the repo and it looks confusing to me.

tretre91 replied:
I added the DDC copyright notice and the tuple notice with a mention that it's the original copyright notice, don't know if that's what you meant

tpadioleau commented on src/ddc/detail/tuple/impl/apply.hpp:
```diff
@@ -0,0 +1,159 @@
+// NOLINTBEGIN(readability-identifier-naming)
+// SPDX-FileCopyrightText: 2026 CExA-project
+// SPDX-License-Identifier: MIT or Apache-2.0 with LLVM-exception
```
Was it originally multi-licensed ?

tretre91 replied:
The tuple (and the other CExA-experimental projects) is multi-licensed MIT and Apache

tpadioleau replied:
I mean the implementation originally comes from an existing project, was it licensed Apache ?

tpadioleau replied:
my mistake, you implemented tuple by yourself.

tpadioleau commented on src/ddc/detail/tuple/impl/apply.hpp:
```diff
@@ -0,0 +1,159 @@
+namespace cexa {
+
+namespace impl {
+template <class>
+constexpr bool is_reference_wrapper_v = false;
+template <class U>
+constexpr bool is_reference_wrapper_v<std::reference_wrapper<U>> = true;
```
```suggestion
```

Can we really support `std::reference_wrapper` in a device compatible tuple ?

tretre91 replied:
You still have the option to only use the tuple on host (but I agree on device this shouldn't be possible)

tpadioleau commented on src/ddc/detail/tuple/impl/helper.hpp:
```diff
@@ -0,0 +1,152 @@
+#include <array>
+#include <cstddef>
+#include <ranges>
+#include <type_traits> // integral_constant
```
```suggestion
#include <type_traits>
```

tpadioleau commented on src/ddc/detail/tuple/impl/apply.hpp:
```diff
@@ -0,0 +1,159 @@
+
+    if constexpr (std::is_function_v<Pointed>) {
+        if constexpr (is_derived_object) {
+            return (std::forward<Object>(object).*member)(std::forward<Args>(args)...);
```
For DDC usage of `std::forward` is ok on device code but in Kokkos I guess it is not ?

tretre91 replied:
`std::forward` in a device function should be fine, [nvcc implicitely treats move and forward as `__host__ __device__`](https://docs.nvidia.com/cuda/cuda-programming-guide/05-appendices/cpp-language-support.html#cuda-std-move-cuda-std-forward) and I didn't have issue on hip (I guess they do the same)
I could find uses of `std::forward` on device in kokkos, here for example https://github.com/kokkos/kokkos/blob/6d75fd77a7e854928ab9a2ee2b20cb80cfb22457/core/src/impl/Kokkos_TeamMDPolicy.hpp#L192-L219

tpadioleau replied:
Nice, I did not know!

tpadioleau commented on src/ddc/detail/tuple/impl/ignore.hpp:
```diff
@@ -0,0 +1,23 @@
+} // namespace impl
+
+// TODO: check if we need a host and a device version of this
+inline constexpr impl::ignore_t ignore;
```
IMO should be ok as long as we don't take a reference or the adress from the device

tpadioleau commented on src/ddc/detail/tuple/impl/macros.hpp:
```diff
@@ -0,0 +1,29 @@
+#endif
+
+#if defined(KOKKOS_COMPILER_NVCC)
+#    define CEXA_NVCC_HOST_DEVICE_CHECK_DISABLE _Pragma("nv_exec_check_disable")
```
Should we check this disabling does not leak in functions where we actually want a diagnostic ?

tretre91 replied:
It only applies to the symbol defined immediately after the pragma. If you construct a tuple with host only types you won't get a warning, but if you construct your tuple with a host device type `T`, and `T`'s constructor calls a host only function you will get a warning about the call in `T` constructor, I don't know if that's clear

I added it because when using the tuple on host with host only types there were too many warnings (which are false positives since we're staying on host)

tpadioleau commented on src/ddc/kernels/splines/spline_evaluator_nd.hpp:
```diff
@@ -0,0 +1,809 @@
+
+#include "ddc/coordinate.hpp"
+#include "ddc/detail/type_seq.hpp"
+#include "ddc/discrete_domain.hpp"
```
```suggestion
```

DDC components only use `<ddc/ddc.hpp>`

tpadioleau commented on src/ddc/kernels/splines/spline_evaluator_nd.hpp:
```diff
@@ -0,0 +1,809 @@
+#include "integrals.hpp"
+#include "periodic_extrapolation_rule.hpp"
+
+namespace ddc {
```
```suggestion
namespace ddc {

```

tpadioleau commented on src/ddc/kernels/splines/spline_evaluator_nd.hpp:
```diff
@@ -0,0 +1,809 @@
+ * - ExtrapolationRule A TypeSeq containing the lower and upper extrapolation rules along each dimension of interest.
+ */
+template <class... Args>
+class SplineEvaluatorND;
```
```suggestion
template <
        class ExecSpace,
        class MemorySpace,
        class BSplines,
        class EvaluationDDims,
        class ExtrapolationRules>
class SplineEvaluatorND;
```

tpadioleau commented on src/ddc/kernels/splines/spline_evaluator_nd.hpp:
```diff
@@ -0,0 +1,809 @@
+// A value that can be used to do a pack expansion over (0, 1, ..., Dimension)
+#define IDX ddc::type_seq_rank_v<BSplines, bsplines_ts>
+
+    using bsplines_ts = detail::TypeSeq<BSplines...>;
```
```suggestion
    using bsplines_ts = detail::TypeSeq<BSplines...>;
    // A value that can be used to do a pack expansion over (0, 1, ..., Dimension)
    template <class BSpline>
    static constexpr std::size_t s_idx = ddc::type_seq_rank_v<BSpline, bsplines_ts>;
```
Should be ok ? Slightly longer to write but avoids a macro.

tpadioleau commented on src/ddc/kernels/splines/spline_evaluator_nd.hpp:
```diff
@@ -0,0 +1,809 @@
+                    auto const spline_coef_ND = spline_coef[j];
+                    ddc::device_for_each(
+                            evaluation_domain(spline_eval.domain()),
+                            [=, *this](evaluation_domain::discrete_element_type const i) {
```
```suggestion
                            [&](evaluation_domain::discrete_element_type const i) {
```

tpadioleau commented on src/ddc/kernels/splines/spline_evaluator_nd.hpp:
```diff
@@ -0,0 +1,809 @@
+
+                    ddc::device_for_each(
+                            evaluation_domain(spline_eval.domain()),
+                            [=, *this](evaluation_domain::discrete_element_type const i) {
```
```suggestion
                            [&](evaluation_domain::discrete_element_type const i) {
```

tpadioleau commented on src/ddc/kernels/splines/spline_evaluator_nd.hpp:
```diff
@@ -0,0 +1,809 @@
+                    auto const spline_coef_ND = spline_coef[j];
+                    ddc::device_for_each(
+                            evaluation_domain(spline_eval.domain()),
+                            [=, *this](evaluation_domain::discrete_element_type const i) {
```
```suggestion
                            [&](evaluation_domain::discrete_element_type const i) {
```

tpadioleau commented on src/ddc/kernels/splines/spline_evaluator_nd.hpp:
```diff
@@ -0,0 +1,809 @@
+                    auto const spline_coef_ND = spline_coef[j];
+                    ddc::device_for_each(
+                            evaluation_domain(spline_eval.domain()),
+                            [=, *this](evaluation_domain::discrete_element_type const i) {
```
```suggestion
                            [&](evaluation_domain::discrete_element_type const i) {
```

tpadioleau commented on src/ddc/kernels/splines/spline_evaluator_nd.hpp:
```diff
@@ -0,0 +1,809 @@
+                    integrals(j) = 0;
+                    ddc::device_for_each(
+                            ddc::DiscreteDomain<BSplines...>(),
+                            [=](ddc::DiscreteDomain<BSplines...>::discrete_element_type const i) {
```
```suggestion
                            [&](ddc::DiscreteDomain<BSplines...>::discrete_element_type const i) {
```

tpadioleau commented on src/ddc/kernels/splines/spline_evaluator_nd.hpp:
I see multiple extended lambdas inside member functions of a class with multiple variadic template parameters, I hope this will work fine for nvcc.

tretre91 replied:
I didn't have any problem so far, but yes the cuda programming guide says this about the template parameter restrictions
> If the enclosing function is an instantiation of a function or member template, or if the function is a member of a class template, then the template(s) must satisfy the following constraints: ...

I can try to only use a single parameter pack

tpadioleau replied:
not needed for now

tpadioleau commented on src/ddc/kernels/splines/spline_evaluator_nd.hpp:
```diff
@@ -0,0 +1,809 @@
+                KOKKOS_LAMBDA(batch_domain_type<BatchedDDom>::discrete_element_type const j) {
+                    integrals(j) = 0;
+                    ddc::device_for_each(
+                            ddc::DiscreteDomain<BSplines...>(),
```
```suggestion
                            ddc::DiscreteDomain<BSplines...>(spline_coef.domain()),
```
I think it would even be better to construct this object outside of the parallel_for_each to avoid each thread paying for this ?

tretre91 replied:
Sure

tpadioleau reviewed (changes requested):
Great job, let me know if some comments are not clear or need discussion

tretre91 committed:
Replace the IDX macro with a template variable

Co-authored-by: Thomas Padioleau <thomas.padioleau@cea.fr>

tretre91 committed:
Capture by reference and construct the domains outside the kernels

tretre91 committed:
Add an entry in the AUTHORS file

tretre91 committed:
Update the tuple copyright statement

tretre91 committed:
fix typo in the AUTHORS file

tpadioleau committed:
Add const

tpadioleau committed:
Format

tpadioleau committed:
Merge branch 'main' into nd_evaluator

tpadioleau commented on src/ddc/detail/tuple/impl/apply.hpp:
```diff
@@ -0,0 +1,165 @@
+// Copyright (C) The DDC development team, see COPYRIGHT.md file
+//
+// SPDX-License-Identifier: MIT
```
Was it the DDC CI that forced you to add this header ?

tretre91 replied:
No, I added it because I thought you were asking for it in https://github.com/CExA-project/ddc/pull/967#discussion_r3218888087

It could be removed, technically since I add a `// NOLINTBEGIN(readability-identifier-naming)` it counts as a modification of the sources so it could make sense to keep it alongside the original copyright notice. I'm fine either way.

tpadioleau replied:
I see, sorry for the confusion, then I think we should just put the headers you have in the cexa repo. I can do it if you prefer.

tretre91 committed:
Only use the original copyright notice for the tuple

tpadioleau committed:
Merge branch 'main' into nd_evaluator
